### PR TITLE
Add GSTA processor (from SimVP2)

### LINKED ISSUE #383: Add SimVP model
### Problem / Goal

Add a SimVP model

### What does good look like?

- [ ] SimVP model can be run

### Related context

- Parent issue: #156

### PR DESCRIPTION
**Pull request by jemrobinson**
## Summary

Adds a GSTAProcessor that implements the **translator** component of SimVP2 [Tan et al., 2022](https://arxiv.org/abs/2211.12509). This uses a flat stack of gSTA (gated spatio-temporal attention) blocks instead of the U-Net of Inception blocks used in SimVPv1. The implementation of GatedAttentionBlock used here is slightly modified to allow different numbers of input and output channels.

**TL;DR** `GSTAProcessor` is faster and more accurate than `ViT` (using the same `CNN` encoder/decoder) and also faster and more accurate than standalone `DDPM`.

Closes #383

## Other changes in this PR

- Fix to `SingleDataset.frequency` to handle incorrect Anemoi metadata when one of the first two dates in a dataset is missing.
- Reduce DPI in W&B images/videos during training

## Model summary

- timesteps are folded into the channel dimension
- the NCHW tensor is passed through a flat stack of `GatedAttentionBlock`s
- each `GatedAttentionBlock` is
  - subblock 1
      - normalise the input
      - apply gated attention to select spatial features
      - apply learnable per-channel scale
      - randomly drop a subset of samples for regularisation
      - add a skip connection to the input
  - subblock 2
      - normalise the input
      - apply a conv-MLP to mix channels
      - apply learnable per-channel scale
      - randomly drop a subset of samples for regularisation
      - add a skip connection to the input
  - a final convolution to set the output channels

<details>
<summary>Processor diagram</summary>

```mermaid
flowchart TD
    X["x"] --> N1["BatchNorm2d"]
    N1 --> ATT["GatedAttention block"]

    ATT --> LS1["layer_scale_1"]
    LS1 --> DP1["DropPath"]
    X --> ADD1(("+"))
    DP1 --> ADD1

    ADD1 --> N2["BatchNorm2d"]
    N2 --> MLPBLK["ConvMLP block"]

    MLPBLK --> LS2["layer_scale_2"]
    LS2 --> DP2["DropPath"]
    ADD1 --> ADD2(("+"))
    DP2 --> ADD2

    ADD2 --> PROJ{"in_channels == out_channels?"}
    PROJ -->|"no"| CONV1["Conv1x1"]
    PROJ -->|"yes"| IDENT["identity"]
    CONV1 --> OUT["output"]
    IDENT --> OUT

    classDef terminal fill:#e8e8e8,stroke:#888,color:#111
    classDef conv fill:#cde4ff,stroke:#4a77bd,color:#111
    classDef attention fill:#ffe4b5,stroke:#cc8400,color:#111
    classDef sample fill:#d5f5d5,stroke:#3a9d3a,color:#111
    classDef optional fill:#f0f0f0,stroke:#999,color:#555,stroke-dasharray:5 5
    classDef drop fill:#f8d0d0,stroke:#c0392b,color:#111
    classDef dots fill:none,stroke:none,color:#777

    class X,OUT terminal
    class ATT attention
    class CONV1,MLPBLK conv
    class LS1,LS2 optional
    class N1,N2 sample
    class DP1,DP2 drop
```
</details>

## Comparison

Ran the following models from the same commit (00830bb):

- Models
  - **cnn-gsta-cnn**: `CNNEncoder` > `GSTAProcessor` > `CNNDecoder`
  - **cnn-vit-cnn**: `CNNEncoder` > `ViTProcessor` > `CNNDecoder`
  - **ddpm**: `DDPM`
- Both hemispheres (north/south)
- Same dates: `n_forecast_steps=12`, `n_history_steps=3`
- Same data: `sample-float-argo`, `sample-sic-ssmis`, `sample-era5` inputs; `sample-sic-ssmis` target
- Same CNN-encoder/decoder
- Same `HuberLoss` / `ModelCheckpoint(monitor=validation_loss)` / `EarlyStopping(patience=20)` across all stages.
- Same architecture (JASMIN, 1 GPU)

### Training

| Model | Hemisphere | Best epoch | Validation loss @ best | Time @ best |
|---|---|---|---|---|
| `cnn_gsta_cnn` | `north`	| **47** | **0.00101** | **5h03** |
| `cnn_vit_cnn` | `north` | 54 | 0.01075 | 7h03 |
| `ddpm` | `north` | 161 | 0.00469 | 10h23 |
| `cnn_gsta_cnn` | `south` | **14** | **0.00092** | **1h31** |
| `cnn_vit_cnn` | `south` | 183 | 0.00389 | 24h (timed out) |
| `ddpm` | `south` | 430 | 0.00274 | 24h (timed out) |

**Total training compute:**
- `cnn-gsta-cnn`: 1-5 hrs
- `cnn-vit-cnn`: 7-24 hrs
- `ddpm`: 10-24 hrs

### North

| Lead | Metric | `cnn-gsta-cnn` | `cnn-vit-cnn` | `ddpm` |
|---|---|---|---|---|
| 1 | accuracy | **99.613** | 97.643 | 96.421 |
| 1 | mae | **0.00649** | 0.03599 | 0.01844 |
| 1 | rmse | **0.02895** | 0.11810 | 0.05927 |
| 1 | sieerror (km2) | **125,708** | 610,000 | 4,082,250 |
| 2 | accuracy | **99.570** | 97.363 | 94.044 |
| 2 | mae | **0.00727** | 0.04693 | 0.02535 |
| 2 | rmse | **0.03264** | 0.14155 | 0.08555 |
| 2 | sieerror (km2) | **60,042** | 1,108,250 | 6,927,750 |

### South

| Lead | Metric | `cnn-gsta-cnn` | `cnn-vit-cnn` | `ddpm` |
|---|---|---|---|---|
| 1 | accuracy | **99.030** | 96.625 | 99.538 |
| 1 | mae | **0.02256** | 0.02315 | 0.02308 |
| 1 | rmse | 0.04979 | 0.09530 | **0.04807** |
| 1 | sieerror (km2) | 703,292 | 2,714,292 | **459,333** |
| 2 | accuracy | **98.828** | 96.646 | 77.165 |
| 2 | mae | 0.02288 | **0.02233** | 0.06171 |
| 2 | rmse | **0.05124** | 0.08861 | 0.12448 |
| 2 | sieerror (km2) | **1,045,042** | 3,102,625 | 26,634,208 |

<details>
<summary><b>Per-forecast-day metrics (north)</b></summary>

sieerror_per_forecast_day_north
rmse_per_forecast_day_north
mae_per_forecast_day_north
accuracy_per_forecast_day_north

</details>

<details>
<summary><b>Per-forecast-day metrics (south)</b></summary>

sieerror_per_forecast_day_south
rmse_per_forecast_day_south
mae_per_forecast_day_south
accuracy_per_forecast_day_south

</details>

### Evaluation plots

<details>
<summary><b>North</b></summary>

2019-01-12-cnn-gsta-cnn-north
2019-01-12-cnn-vit-cnn-north
2019-01-12-ddpm

</details>

<details>
<summary><b>South</b></summary>

2019-01-12-cnn-gsta-cnn-south
2019-01-12-cnn-vit-cnn-south
2019-01-12-dppm-south

</details>

## Supporting information

<details>
<summary><b>W&B run links</b></summary>

| Run | Purpose | Link |
|---|---|---|
| cerulean-galaxy-1361 | `cnn_gsta_cnn (north)` training  | https://wandb.ai/turing-seaice/train/runs/1ytaha7j |
| rural-salad-384 | `cnn_gsta_cnn (north)` evaluate | https://wandb.ai/turing-seaice/evaluate/runs/puo4ayro |
| fine-oath-1362 | `cnn_vit_cnn (north)` training | https://wandb.ai/turing-seaice/train/runs/ib35k0p4 |
| restful-breeze-385 | `cnn_vit_cnn (north)` evaluate | https://wandb.ai/turing-seaice/evaluate/runs/sarsassr |
| worldly-microwave-1359 | `ddpm (north)` training | https://wandb.ai/turing-seaice/train/runs/cc1501ob |
| visionary-gorge-387 | `ddpm (north)` evaluate | https://wandb.ai/turing-seaice/evaluate/runs/wfqyeuss |
| fresh-forest-1360 | `cnn_gsta_cnn (south)` training | https://wandb.ai/turing-seaice/train/runs/unrjivdm |
| fearless-lake-389 | `cnn_gsta_cnn (south)` evaluate | https://wandb.ai/turing-seaice/evaluate/runs/eh7gc522 |
| feasible-grass-1357 | `cnn_vit_cnn (south)` training | https://wandb.ai/turing-seaice/train/runs/55w96jlx |
| fanciful-sea-386 | `cnn_vit_cnn (south)` evaluate | https://wandb.ai/turing-seaice/evaluate/runs/2130waad |
| hopeful-river-1358 | `ddpm (south)` training | https://wandb.ai/turing-seaice/train/runs/qoe3dac1 |
| eager-star-388 | `ddpm (south)` evaluate | https://wandb.ai/turing-seaice/evaluate/runs/l20lh7t0 |
</details>

### CODE REVIEW

**Comment by marianovitasari20** [comment]
reviewing this

File: icenet_mp/models/common/gated_attention.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    - a depthwise conv (kernel 2*dilation-1)
+    - a depthwise dilated conv (kernel ~= kernel_size // dilation)
+    - a gated 1x1 projection.
+    """
+
+    def __init__(self, channels: int, *, kernel_size: int, dilation: int) -> None:
+        """Initialise a GatedAttention module."""
+        super().__init__()
+        depthwise_kernel = 2 * dilation - 1
+        dilated_kernel = kernel_size // dilation + ((kernel_size // dilation) % 2 - 1)
```
Comment (marianovitasari20) [inline icenet_mp/models/common/gated_attention.py:24]:
```suggestion
        dilated_kernel = max(kernel_size // dilation + ((kernel_size // dilation) % 2 - 1), 1)
```

Comment (marianovitasari20) [reply icenet_mp/models/common/gated_attention.py:24]: If `kernel_size < dilation`, this can produce a negative kernel size, which is not a valid input to `nn.Conv2d`. For example, `kernel_size // dilation` evaluates to 0, and the expression becomes `0 + (0 % 2 - 1) = -1`.

A minimal fix would be to just clamp the result to a minimum of 1:
```python
dilated_kernel = max(kernel_size // dilation + ((kernel_size // dilation) % 2 - 1), 1)

```

but feel free to take a different approach.

**Review by marianovitasari20 — CHANGES_REQUESTED** [review]

File: icenet_mp/models/common/gated_attention.py (hunk 1 above)
Comment (marianovitasari20) [reply icenet_mp/models/common/gated_attention.py:24]: Given this edge case, would it be worth adding a test that constructs a `GatedAttentionBlock` with `kernel_size < dilation` to confirm the fix actually prevents the crash?

Comment (jemrobinson) [reply icenet_mp/models/common/gated_attention.py:24]: Done in 57a4515a

File: icenet_mp/models/common/gated_attention.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            else nn.Conv2d(in_channels, out_channels, kernel_size=1)
+        )
+
+    def _drop_path(self, x: Tensor) -> Tensor:
+        """Stochastic depth: zero this residual branch for a random subset of samples."""
+        if self.drop_path_prob == 0.0 or not self.training:
+            return x
+        keep_prob = 1 - self.drop_path_prob
+        mask = x.new_empty(x.shape[0], *([1] * (x.ndim - 1))).bernoulli_(keep_prob)
+        return x * mask / keep_prob
```
Comment (marianovitasari20) [inline icenet_mp/models/common/gated_attention.py:130]: When `drop_path_prob == 1.0`, `keep_prob` becomes 0.0, and `x * mask / keep_prob` divides by zero. This produces `NaN/inf` silently instead of raising an error. Suggest adding `assert 0.0 <= drop_path_prob < 1.0` in `GatedAttentionBlock` so a bad config value fails immediately at setup rather than deep inside a training step.

Worth a unit test covering both `drop_path_prob=0.0` and `drop_path_prob=1.0` (should now raise, rather than silently produce NaN).

File: icenet_mp/models/processors/gsta.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            hid_channels: Hidden channel count for the temporal translator.
+            kernel_size: Kernel size for gated attention blocks in the translator.
+            mlp_drop_prob: Element-wise dropout probability for the translator conv-MLP blocks.
+            mlp_ratio: Hidden-channel expansion ratio for the translator conv-MLP blocks.
+            n_blocks: Number of gated attention blocks in the temporal translator.
+            kwargs: Arguments to BaseProcessor.
+
+        """
+        super().__init__(**kwargs)
+        if n_blocks < 2:  # noqa: PLR2004
```
Comment (marianovitasari20) [inline icenet_mp/models/processors/gsta.py]: Why is `2` the enforced minimum here? You may have already mentioned this, but it might be worth adding a comment in the code to explain the reasoning.

File: icenet_mp/models/common/gated_attention.py (hunk 2 above)
Comment (jemrobinson) [reply icenet_mp/models/common/gated_attention.py:130]: I think `drop_path_prob=1.0` should return zeros, which is what is now done in d38f6126. Do you agree?

File: icenet_mp/models/processors/gsta.py (hunk 3 above)
Comment (jemrobinson) [reply icenet_mp/models/processors/gsta.py]: I guess it doesn't have to be. If `n_blocks==1` then you go straight from input channels to output channels without hidden channels, but this could be a warning not an error.

Comment (jemrobinson) [reply icenet_mp/models/processors/gsta.py]: Done in 19b5f056

**Review by marianovitasari20 — APPROVED** [review]
LGTM
